### DISCUSSION TITLE
How to inject dependencies in headless xunit tests?

### DISCUSSION THREAD

**Discussion by ClassyCircuit** [AvaloniaUI/Avalonia#14759] [Q&A] [answered]
I am using headless platform for writing end to end tests with xunit test runner and I would like to avoid simulating mouse clicks, but I want to access viewmodels and change properties directly on them.
In the real application I am using MS dependency injection and it is working fine, but I am having a lot of trouble setting it up in the headless project.
If I try to inject any viewmodel into xunit test class, it fails with this error:
`Xunit.Sdk.TestClassException: The following constructor parameters did not have matching fixture data: ExplorerVm explorerVm`

I tried adding `Xunit.DependencyInjection` nuget package and setting up DI container for xunit, but then it looks like I end up with two DI providers and they are clashing.

I don't understand, how do I inject my classes into xunit headless tests and interact with them directly.

**Comment by maxkatz6** [comment]
Not sure if I understand your question.
Usually, you would create/mock IoC container and pass it to your view models. You don't need to use unit tests constructors to pass any data - dependency injection into unit tests isn't really typical solution.

**Reply by ClassyCircuit** [reply]
I am doing integration tests and I don't want to mock any part of it. I want my app to be as close as possible to the real thing. 
My viewmodels and other parts of application heavily use dependency injection, this is how I have setup the real app:

 ```.cs
public partial class App : Application
{
    public IServiceProvider Services { get; set; } = default!;
    public new static App Current => (App)Application.Current!;

    public override void Initialize()
    {
        Services = ConfigureServices();
        AvaloniaXamlLoader.Load(this);
    }

    /// <summary>
    /// Configures the services for the application.
    /// </summary>
    private static IServiceProvider ConfigureServices()
    {
        IServiceCollection services = new ServiceCollection();
        AddUIProject(services);
        return services.BuildServiceProvider();
    }

    /// <summary>
    /// This wires all projects
    /// </summary>
    /// <param name="services"></param>
    public static IServiceCollection AddUIProject(IServiceCollection services)
    {
        services.AddApiProject();
        services.AddDomainProject();
        services.AddInfrastructureProject();

        services.AddMediatR(t => t.RegisterServicesFromAssemblies(
        [
            typeof(UI.App).Assembly,
            typeof(Api.StartupExtensions).Assembly,
            typeof(Domain.StartupExtensions).Assembly,
            typeof(Infrastructure.StartupExtensions).Assembly,
        ]));

        // ViewModels
        services.AddTransient<MainVm>();
        services.AddTransient<SidebarVm>();
        services.AddTransient<ExplorerVm>();
        services.AddTransient<SourceControlVm>();
        services.AddTransient<EditorContainerVm>();

        // Editor ViewModels
        services.AddTransient<RequestEditorView>();
        services.AddTransient<FolderEditorView>();
        services.AddTransient<TextEditorView>();

        // Property editor ViewModels
        services.AddTransient<QueryEditorView>();
        services.AddTransient<HeadersEditorView>();
        services.AddTransient<BodyEditorView>();
        services.AddTransient<AuthEditorView>();

        services.AddTransient<EditorFactory>();
        services.AddTransient<PropertyFactory>();

        return services;
    }
}
```

In unit tests, I don't use DI container, I just test classes/methods directly, but in this case I am trying to do integration tests without any mocks.

And this is how my integration test project is setup:

```.cs
public partial class TestApp : Application
{
    public IServiceProvider Services { get; private set; } = default!;
    public new static TestApp Current => (TestApp)Application.Current!;

    public override void Initialize()
    {
        Services = ConfigureServices();
        AvaloniaXamlLoader.Load(this);
    }

    /// <summary>
    /// Configures the services for the application.
    /// </summary>
    private static IServiceProvider ConfigureServices()
    {
        IServiceCollection services = new ServiceCollection();
        AddUIProject(services);
        return services.BuildServiceProvider();
    }

    /// <summary>
    /// This wires all projects
    /// </summary>
    /// <param name="services"></param>
    public static void AddUIProject(IServiceCollection services)
    {
        UI.App.AddUIProject(services);
    }
}

[assembly: AvaloniaTestApplication(typeof(TestAppBuilder))]

namespace Tests.EndToEnd
{
    public class TestAppBuilder
    {
        public static AppBuilder BuildAvaloniaApp() => AppBuilder.Configure<TestApp>()
            .UseHeadless(new AvaloniaHeadlessPlatformOptions());
    }
}

```

This is a sample integration test where I try to inject a dependency:

```.cs
    public class ExplorerIntegrationTests
    {
        private readonly ExplorerVm _explorerVm;

        
        public ExplorerIntegrationTests(ExplorerVm explorerVm)
        {
            _explorerVm = explorerVm;
        }

        [AvaloniaFact]
        public void ExplorerVm_OnSelectedNodeChanged_Event_Is_Fired()
        {
              // This throws exception: Xunit.Sdk.TestClassException: The following constructor parameters did not have matching fixture data: ExplorerVm explorerVm
            // I would like to access the injected ExplorerVm here, but I don't understand how to do that.
           // I can't just instantiate it here manually, because then I would have to instantiate the whole graph of dependencies manually, that's why I have DI container setup.

        }
    }

```

In other projects, such as ASP .NET Core, I would use Xunit.DependencyInjection package and setup a DI container just for the test project and it would create the whole dependency graph automatically, I would be able to do real end to end tests with infrastructure, but I don't have enough knowledge how Avalonia works to set it up properly.


Is it even possible to run headless platform and interact with viewmodel properties/methods directly without simulating UI clicks?

**Reply by timunie** [reply]
Does this help by any chance? https://github.com/AvaloniaUI/Avalonia.Samples?tab=readme-ov-file#%EF%B8%8F-automated-ui-testing

**Reply by maxkatz6** [reply, marked as answer]
> I want my app to be as close as possible to the real thing.

But why do you resolve `ExplorerVm explorerVm` from the ctor of the test? And why do you expect it to be resolved there from your service provider? There is no connection between them.

Inside of your test method, you should be able to access `TestApp.Current.Services.GetService()`

**Reply by ClassyCircuit** [reply]
@maxkatz6 Hmm, I think you're right, I should stop injecting them and just resolve using the TestApp.Current.Services provider.

> And why do you expect it to be resolved there from your service provider? There is no connection between them.  

To be honest, I don't know, I don't understand how Avalonia headless and xunit startup works together.
